Make headers more self-contained

Some of the Glucat headers are not self-contained. For example, compiling the following code in `framed_multi.cpp`:
```
#include "glucat/framed_multi.h"
int main(int argc, char ** argv)
{ return 0; }
```
using
```
g++ -std=c++11 -std=c++11 -DHAVE_CONFIG_H -I. -I..  -I. -I..     -Wall -g -O3  -march=native -finline-limit=8000 --param large-function-insns=8000 --param max-inline-insns-auto=8000 --param max-inline-insns-single=8000 --param inline-unit-growth=200 --param large-function-growth=200 -funroll-loops -fvariable-expansion-in-unroller -DNDEBUG -D_GLUCAT_USE_STD_UNORDERED_MAP -D_GLUCAT_USE_DENSE_MATRICES -D_GLUCAT_CHECK_ISNAN -Wno-unused-local-typedefs -Wno-deprecated-register -fno-check-new -DHAVE_BITS_WORDSIZE_H -D_GLUCAT_HAVE_BOOST_SERIALIZATION_ARRAY_WRAPPER_H -fexceptions -MT framed_multi.o -MD -MP -MF .deps/framed_multi.Tpo -c -o framed_multi.o framed_multi.cpp
```
results in
```
In file included from framed_multi.cpp:32:
../glucat/framed_multi.h:118:56: error: ISO C++ forbids declaration of ‘type name’ with no type [-fpermissive]
  118 |   operator<< (std::ostream& os, const std::pair< const index_set<LO,HI>, Scalar_T >& term) -> std::ostream&;
      |                                                        ^~~~~~~~~
../glucat/framed_multi.h:118:83: error: wrong number of template arguments (1, should be 2)
  118 |   operator<< (std::ostream& os, const std::pair< const index_set<LO,HI>, Scalar_T >& term) -> std::ostream&;
      |                                                                                   ^
In file included from /usr/include/c++/11/bits/stl_algobase.h:64,
                 from /usr/include/c++/11/bits/char_traits.h:39,
                 from /usr/include/c++/11/string:40,
                 from ../glucat/errors.h:34,
                 from ../glucat/framed_multi.h:35,
                 from framed_multi.cpp:32:
/usr/include/c++/11/bits/stl_pair.h:211:12: note: provided for ‘template<class _T1, class _T2> struct std::pair’
  211 |     struct pair
      |            ^~~~
[...]
```
See (e.g.) https://www.modernescpp.com/index.php/c-core-guidelines-the-remaining-rules-to-source-files

For now, the headers that use `Tune_P` will still need to depend on the user or tests defining this type. This will be treated in a separate issue.

--- a/glucat/clifford_algebra.h
+++ b/glucat/clifford_algebra.h
@@ -31,6 +31,8 @@
      See also Arvind Raja's original header comments in glucat.h
  ***************************************************************************/
 
+#include "glucat/global.h"
+
 #include <string>
 #include <utility>
 #include <vector>

--- a/glucat/clifford_algebra_imp.h
+++ b/glucat/clifford_algebra_imp.h
@@ -54,9 +54,9 @@
 // Doron Zeilberger, "PADE" (Maple code), 2002.
 // http://www.math.rutgers.edu/~zeilberg/tokhniot/PADE
 
-#include <glucat/clifford_algebra.h>
+#include "glucat/clifford_algebra.h"
 
-#include <glucat/scalar.h>
+#include "glucat/scalar.h"
 
 #include <array>
 

--- a/glucat/errors_imp.h
+++ b/glucat/errors_imp.h
@@ -31,6 +31,12 @@
      See also Arvind Raja's original header comments in glucat.h
  ***************************************************************************/
 
+#include "glucat/errors.h"
+
+#include <string>
+#include <iostream>
+#include <ostream>
+
 namespace glucat
 {
   /// Specific exception class

--- a/glucat/framed_multi.h
+++ b/glucat/framed_multi.h
@@ -33,6 +33,7 @@
 
 #include "glucat/global.h"
 #include "glucat/errors.h"
+#include "glucat/index_set.h"
 #include "glucat/clifford_algebra.h"
 
 #if defined(_GLUCAT_USE_BOOST_POOL_ALLOC)

--- a/glucat/framed_multi_imp.h
+++ b/glucat/framed_multi_imp.h
@@ -34,7 +34,10 @@
 
 #include "glucat/framed_multi.h"
 
+#include "glucat/scalar.h"
 #include "glucat/random.h"
+#include "glucat/generation.h"
+#include "glucat/matrix.h"
 
 #if defined(_GLUCAT_USE_BOOST_POOL_ALLOC)
 // Use the Boost pool allocator

--- a/glucat/generation.h
+++ b/glucat/generation.h
@@ -31,6 +31,10 @@
      See also Arvind Raja's original header comments in glucat.h
  ***************************************************************************/
 
+#include "glucat/global.h"
+
+#include <boost/numeric/ublas/fwd.hpp>
+
 #include <utility>
 #include <array>
 #include <map>

--- a/glucat/generation_imp.h
+++ b/glucat/generation_imp.h
@@ -31,8 +31,8 @@
      See also Arvind Raja's original header comments in glucat.h
  ***************************************************************************/
 
+#include "glucat/global.h"
 #include "glucat/generation.h"
-
 #include "glucat/matrix.h"
 
 namespace glucat { namespace gen

--- a/glucat/glucat.h
+++ b/glucat/glucat.h
@@ -53,6 +53,8 @@
 
 #include "glucat/qd.h"
 
+#include "glucat/random.h"
+
 #include "glucat/clifford_algebra.h"
 
 #include "glucat/framed_multi.h"

--- a/glucat/index_set_imp.h
+++ b/glucat/index_set_imp.h
@@ -33,6 +33,7 @@
 
 #include "glucat/index_set.h"
 
+#include <string>
 #include <sstream>
 
 namespace glucat

--- a/glucat/long_double.h
+++ b/glucat/long_double.h
@@ -31,6 +31,9 @@
  See also Arvind Raja's original header comments and references in glucat.h
  ***************************************************************************/
 
+#include "glucat/global.h"
+#include "glucat/scalar.h"
+
 namespace glucat
 {
 #if defined(__USE_GNU)

--- a/glucat/matrix_imp.h
+++ b/glucat/matrix_imp.h
@@ -32,6 +32,8 @@
      See also Arvind Raja's original header comments in glucat.h
  ***************************************************************************/
 
+#include "glucat/errors.h"
+#include "glucat/scalar.h"
 #include "glucat/matrix.h"
 
 # if  defined(_GLUCAT_GCC_IGNORE_UNUSED_LOCAL_TYPEDEFS)

--- a/glucat/matrix_multi_imp.h
+++ b/glucat/matrix_multi_imp.h
@@ -33,8 +33,9 @@
 
 #include "glucat/matrix_multi.h"
 
-#include "glucat/matrix.h"
+#include "glucat/scalar.h"
 #include "glucat/generation.h"
+#include "glucat/matrix.h"
 
 # if  defined(_GLUCAT_GCC_IGNORE_UNUSED_LOCAL_TYPEDEFS)
 #  pragma GCC diagnostic push

--- a/glucat/qd.h
+++ b/glucat/qd.h
@@ -31,6 +31,7 @@
  See also Arvind Raja's original header comments and references in glucat.h
  ***************************************************************************/
 
+#include "glucat/global.h"
 #include "glucat/scalar.h"
 
 #if defined(_GLUCAT_USE_QD)

--- a/glucat/scalar.h
+++ b/glucat/scalar.h
@@ -32,6 +32,7 @@
  ***************************************************************************/
 
 #include "glucat/portability.h"
+#include "glucat/global.h"
 
 #include <boost/numeric/ublas/traits.hpp>
 

--- a/glucat/scalar_imp.h
+++ b/glucat/scalar_imp.h
@@ -31,7 +31,8 @@
  See also Arvind Raja's original header comments and references in glucat.h
  ***************************************************************************/
 
-#include "glucat/portability.h"
+#include "glucat/scalar.h"
+#include "glucat/qd.h"
 
 #include <boost/numeric/ublas/traits.hpp>
 

